src/Compilers/CSharp/Portable/Symbols/Metadata/PE/TupleTypeDecoder.cs

```diff
@@ -0,0 +1,313 @@
+                    // We may have a tuple type from a substituted type symbol,
+                    // but it will be missing names from metadata, so we'll
+                    // need to re-create the type.
+                    return type.IsTupleType
+                        ? DecodeNamedType(type.TupleUnderlyingType)
+                        : DecodeNamedType((NamedTypeSymbol)type);
+
+                case SymbolKind.ArrayType:
+                    return DecodeArrayType((ArrayTypeSymbol)type);
+
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(type.TypeKind);
+            }
+        }
+
+        private NamedTypeSymbol DecodeNamedType(NamedTypeSymbol type)
+        {
+            // First decode the type arguments
+            var typeArgs = type.TypeArgumentsNoUseSiteDiagnostics;
+            var decodedArgs = DecodeTypeArguments(typeArgs);
+
+            NamedTypeSymbol decodedType = type;
+
+            // Now check the container
+            NamedTypeSymbol containingType = type.ContainingType;
+            NamedTypeSymbol decodedContainingType;
+            if ((object)containingType != null && containingType.IsGenericType)
+            {
+                decodedContainingType = DecodeNamedType(containingType);
+                Debug.Assert(decodedContainingType.IsGenericType);
+            }
+            else
+            {
+                decodedContainingType = containingType;
+            }
+
+            // Replace the type if necessary
+            var containerChanged = !ReferenceEquals(decodedContainingType, containingType);
+            var typeArgsChanged = typeArgs != decodedArgs;
+            if (typeArgsChanged || containerChanged)
+            {
+                var newTypeArgs = type.HasTypeArgumentsCustomModifiers
+                    ? decodedArgs.Zip(type.TypeArgumentsCustomModifiers,
+                                      (t, m) => new TypeWithModifiers(t, m)).AsImmutable()
+                    : decodedArgs.SelectAsArray(TypeMap.TypeSymbolAsTypeWithModifiers);
+
+                if (containerChanged)
+                {
+                    decodedType = decodedType.OriginalDefinition.AsMember(decodedContainingType);
+                    return decodedType.ConstructIfGeneric(newTypeArgs);
+                }
+
+                decodedType = type.ConstructedFrom.Construct(newTypeArgs, unbound: false);
+            }
+
+            // Now decode into a tuple, if it is one
+            int tupleCardinality;
+            if (decodedType.IsTupleCompatible(out tupleCardinality))
+            {
+                var elementNames = EatElementNamesIfAvailable(tupleCardinality);
+
+                Debug.Assert(elementNames.IsDefault || elementNames.Length == tupleCardinality);
+
+                decodedType = elementNames.IsDefault
+                    ? TupleTypeSymbol.Create(decodedType)
+                    : TupleTypeSymbol.Create(decodedType, elementNames);
```

Comment:
`TupleTypeSymbol.Create()` should allow `elementNames` to be `default(ImmutableArray<string>)`.